### ISSUE TITLE
Inconsistent use of type alias.

### ISSUE THREAD

**Issue by DodoMomo** [nlohmann/json#1507]
Inside `basic_json` the alias `reference` and `const_reference` are used in inconsistently, for example the method `iteration_proxy` uses `const_reference` as a parameter type, while `push_back` uses `const basic_json&`.
Which one should be used? I think the non-alias are better and i would like make a PR if you agree.

Sorry about my horrible english.

**Comment by nlohmann** [comment]
That is indeed inconsistent. Thanks for reporting. A pull request would be more than welcome!

**Comment by garethsb-sony** [comment]
Note that e.g. ``std::vector`` defines ``reference`` but uses ``const value_type&`` in ``push_back``. I've always assumed the rationale was a reference to a stored value vs an input arg.

**Comment by nlohmann** [comment]
I just had a look at the source of `<vector>`, and it seems they use `reference` and `const_reference` internally:

```cpp
const_reference
operator[](size_type __n) const
{ return *(this->_M_impl._M_start + __n); }
```

I also tried to do this in `basic_json`, and I understood this issue to have found spots where I used `const basic_json&` instead of `const_reference` somewhere. Looking at #1523, it seems that @DodoMomo wants to go the other way and replace all `reference´/`const_reference`. I am not sure that this is a good idea.
